infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -365,7 +372,33 @@ private void createIncomingSessionsBySubscriber(ReplicationSubscriber subscriber
         incomingSessions.addAll(sessionsToAdd);
         log.info("Total of {} incoming sessions created with subscriber {}, sessions={}", sessionsToAdd.size(),
                 subscriber, sessionsToAdd);
+        configManager.generateConfig(sessionsToAdd, true);
+    }
 
+    public synchronized void addSessionForClient(LogReplicationSession sessionFromSource) {
+        Set<LogReplicationSession> sessionsToAdd = new HashSet<>();
+        try {
+            IRetry.build(IntervalRetry.class, () -> {
+                try (TxnContext txn = corfuStore.txn(LogReplicationMetadataManager.NAMESPACE)) {
+                    if (!sessions.contains(sessionFromSource)) {
+                        sessionsToAdd.add(sessionFromSource);
+                        metadataManager.addSession(txn, sessionFromSource, topology.getTopologyConfigId(), true);
+                    }
+                    txn.commit();
+                    return null;
+                } catch (TransactionAbortedException e) {
+                    log.error("Failed to create sessions. Retrying.", e);
+                    sessionsToAdd.clear();
+                    throw new RetryNeededException();
+                }
+            }).run();
+        } catch (InterruptedException e) {
+            log.error("Unrecoverable Corfu Error when creating the sessions", e);
+            throw new UnrecoverableCorfuInterruptedError(e);
+        }
+        newSessionsDiscovered.addAll(sessionsToAdd);
+        sessions.addAll(sessionsToAdd);
+        incomingSessions.addAll(sessionsToAdd);
```

Comment:
The snippet above seems redundant. I think the code can be moved or reorganized a bit to avoid code duplication